From 3009947d5e9f47e1122b256f27187a42890446fa Mon Sep 17 00:00:00 2001
From: Michal Kubecek <mkubecek@suse.cz>
Date: Sat, 9 Dec 2017 00:08:21 +0100
Subject: [PATCH 12/19] netlink: add netlink handler for geee (--show-eee)

Implement "ethtool --show-eee <dev>" subcommand using netlink interface
command ETHTOOL_CMD_GET_PARAMS with ETH_PARAMS_IM_EEE mask.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 ethtool.c |  3 ++-
 netlink.c | 68 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 netlink.h |  1 +
 3 files changed, 71 insertions(+), 1 deletion(-)

diff --git a/ethtool.c b/ethtool.c
index a737b3b10be1..ef31c621d12d 100644
--- a/ethtool.c
+++ b/ethtool.c
@@ -4611,6 +4611,7 @@ int nl_gcoalesce(struct cmd_context *ctx) { return 0; }
 int nl_gring(struct cmd_context *ctx) { return 0; }
 int nl_gpause(struct cmd_context *ctx) { return 0; }
 int nl_gchannels(struct cmd_context *ctx) { return 0; }
+int nl_geee(struct cmd_context *ctx) { return 0; }
 #endif
 
 static const struct option {
@@ -4778,7 +4779,7 @@ static const struct option {
 	  "		[ hex on|off ]\n"
 	  "		[ offset N ]\n"
 	  "		[ length N ]\n" },
-	{ "--show-eee", 1, do_geee, NULL,
+	{ "--show-eee", 1, do_geee, nl_geee,
 	  "Show EEE settings"},
 	{ "--set-eee", 1, do_seee, NULL,
 	  "Set EEE settings",
diff --git a/netlink.c b/netlink.c
index 5f58f5a6c59d..03ec6223c0d3 100644
--- a/netlink.c
+++ b/netlink.c
@@ -1756,6 +1756,58 @@ static int show_channels(struct nl_context *nlctx, const struct nlattr *nest)
 	return 0;
 }
 
+static int show_eee(struct nl_context *nlctx, const struct nlattr *nest)
+{
+	const struct nlattr *tb[ETHA_EEE_MAX + 1] = {};
+	DECLARE_ATTR_TB_INFO(tb);
+	bool active, enabled, tx_lpi_enabled;
+	int ret;
+
+	if (!nest)
+		return -EOPNOTSUPP;
+	ret = mnl_attr_parse_nested(nest, attr_cb, &tb_info);
+	if (ret < 0)
+		return ret;
+	if (!tb[ETHA_EEE_LINK_MODES] || !tb[ETHA_EEE_PEER_MODES] ||
+	    !tb[ETHA_EEE_ACTIVE] || !tb[ETHA_EEE_ENABLED] ||
+	    !tb[ETHA_EEE_TX_LPI_ENABLED] || !tb[ETHA_EEE_TX_LPI_TIMER])
+		return -EFAULT;
+
+
+	active = mnl_attr_get_u8(tb[ETHA_EEE_ACTIVE]);
+	enabled = mnl_attr_get_u8(tb[ETHA_EEE_ENABLED]);
+	tx_lpi_enabled = mnl_attr_get_u8(tb[ETHA_EEE_TX_LPI_ENABLED]);
+
+	printf("EEE Settings for %s:\n", nlctx->enlhdr->ifname);
+	printf("\tEEE status: ");
+	if (bitset_is_empty(tb[ETHA_EEE_LINK_MODES], true, &ret)) {
+		printf("not supported\n");
+		return 0;
+	}
+	if (!enabled)
+		printf("disabled\n");
+	else
+		printf("enabled - %s\n", active ? "active" : "inactive");
+	printf("\tTx LPI: ");
+	if (tx_lpi_enabled)
+		printf("%u (us)\n",
+		       mnl_attr_get_u32(tb[ETHA_EEE_TX_LPI_TIMER]));
+	else
+		printf("disabled\n");
+
+	ret = dump_link_modes(tb[ETHA_EEE_LINK_MODES], true, LM_CLASS_REAL,
+			      "Supported EEE link modes:  ", NULL, "\n",
+			      "Not reported");
+	ret = dump_link_modes(tb[ETHA_EEE_LINK_MODES], false, LM_CLASS_REAL,
+			      "Advertised EEE link modes:  ", NULL, "\n",
+			      "Not reported");
+	ret = dump_link_modes(tb[ETHA_EEE_PEER_MODES], false, LM_CLASS_REAL,
+			      "Link partner advertised EEE link modes:  ", NULL,
+			      "\n", "Not reported");
+
+	return 0;
+}
+
 static int get_params_reply_cb(const struct nlmsghdr *nlhdr, void *data)
 {
 	const struct nlattr *tb[ETHA_PARAMS_MAX + 1] = {};
@@ -1805,6 +1857,15 @@ static int get_params_reply_cb(const struct nlmsghdr *nlhdr, void *data)
 			return MNL_CB_ERROR;
 		}
 	}
+	if (!req_mask || req_mask & ETH_PARAMS_IM_EEE) {
+		ret = show_eee(nlctx, tb[ETHA_PARAMS_EEE]);
+		if ((ret < 0) && !(req_mask & ~ETH_PARAMS_IM_EEE)) {
+			nlctx->exit_code = 1;
+			errno = -ret;
+			perror("Cannot get device EEE settings");
+			return MNL_CB_ERROR;
+		}
+	}
 
 	return MNL_CB_OK;
 }
@@ -1836,3 +1897,10 @@ int nl_gchannels(struct cmd_context *ctx)
 				     get_params_reply_cb,
 				     ETH_PARAMS_IM_CHANNELS, 0);
 }
+
+int nl_geee(struct cmd_context *ctx)
+{
+	return ethnl_std_get_request(ctx, ETHTOOL_CMD_GET_PARAMS,
+				     get_params_reply_cb,
+				     ETH_PARAMS_IM_EEE, 0);
+}
diff --git a/netlink.h b/netlink.h
index 3e8593621671..3163dcf36c37 100644
--- a/netlink.h
+++ b/netlink.h
@@ -14,5 +14,6 @@ int nl_gcoalesce(struct cmd_context *ctx);
 int nl_gring(struct cmd_context *ctx);
 int nl_gpause(struct cmd_context *ctx);
 int nl_gchannels(struct cmd_context *ctx);
+int nl_geee(struct cmd_context *ctx);
 
 #endif /* ETHTOOL_NETLINK_H__ */
-- 
2.15.1

